c/jwk.c

```diff
@@ -0,0 +1,398 @@
+#include "json.h"
+#include "httpserver.h"
+#include "charsets.h"
+#include "httpclient.h"
+#include "zssLogging.h"
+#include "jwt.h"
+#include "jwk.h"
+
+static Json *receiveResponse(ShortLivedHeap *slh, HttpClientContext *httpClientContext, HttpClientSession *session, int *statusOut);
+static Json *doRequest(ShortLivedHeap *slh, HttpClientSettings *clientSettings, TlsEnvironment *tlsEnv, char *path, int *statusOut);
+static void getPublicKey(Json *jwk, x509_public_key_info *publicKeyOut, int *statusOut);
+static int getJwk(JwkContext *context);
+static int checkJwtSignature(JwsAlgorithm algorithm, int sigLen, const uint8_t *signature, int msgLen, const uint8_t *message, void *userData);
+static int decodeBase64Url(const char *data, char *resultBuf, int *statusOut);
+static int jwkTaskMain(RLETask *task);
+
+void configureJwt(HttpServer *server, JwkSettings *settings) {
+  int rc = 0;
+
+  if (!settings) {
+    zowelog(NULL, LOG_COMP_ID_JWK, ZOWE_LOG_DEBUG, "JWT disabled or APIML is not configured\n");
+    return;
+  }
+
+  JwkContext *context = (JwkContext*)safeMalloc(sizeof(*context), "Jwk Context");
+  if (!context) {
+    zowelog(NULL, LOG_COMP_ID_JWK, ZOWE_LOG_DEBUG, "failed to allocate JWK context\n");
+    return;
+  }
+
+  context->settings = settings;
+  if (httpServerInitJwtContextCustom(server, settings->fallback, checkJwtSignature, context, &rc) != 0) {
+    zowelog(NULL, LOG_COMP_ID_JWK, ZOWE_LOG_DEBUG, "failed to init JWT context for HTTP server, rc = %d\n", rc);
+    safeFree((char*)context, sizeof(*context));
+    return;
+  };
+
+  RLETask *task = makeRLETask(server->base->rleAnchor, RLE_TASK_TCB_CAPABLE | RLE_TASK_DISPOSABLE, jwkTaskMain);
+  if (!task) {
+    zowelog(NULL, LOG_COMP_ID_JWK, ZOWE_LOG_DEBUG, "failed to create background task for JWK\n");
+    safeFree((char*)context, sizeof(*context));
+    return;
+  }
+  task->userPointer = context;
+  startRLETask(task, NULL);
+
+  zowelog(NULL, LOG_COMP_ID_JWK, ZOWE_LOG_INFO, ZSS_LOG_JWK_URL_MSG, settings->host, settings->port, settings->path);
+}
+
+static int jwkTaskMain(RLETask *task) {
+  JwkContext *context = (JwkContext*)task->userPointer;
+  JwkSettings *settings = context->settings;
+  const int maxAttempts = 1000;
+  const int retryIntervalSeconds = 30;
+  bool success = false;
+
+  for (int i = 0; i < maxAttempts; i++) {
+    int status = getJwk(context);
+    if (status == JWK_STATUS_OK) {
+      success = true;
+      context->isPublicKeyInitialized = true;
+      break;
+    } else if (status == JWK_STATUS_UNRECOGNIZED_FMT_ERROR) {
+      zowelog(NULL, LOG_COMP_ID_JWK, ZOWE_LOG_WARNING, ZSS_LOG_JWK_UNRECOGNIZED_MSG);
+      break;
+    } else if (status == JWK_STATUS_HTTP_CONTEXT_ERROR) {
+      zowelog(NULL, LOG_COMP_ID_JWK, ZOWE_LOG_DEBUG, "failed to init http context\n");
+      break;
+    } else {
+      zowelog(NULL, LOG_COMP_ID_JWK, ZOWE_LOG_WARNING, ZSS_LOG_JWK_RETRY_MSG,
+              jwkGetStrStatus(status), retryIntervalSeconds);
```

Comment:
Thanks,  `JWK_STATUS_PUBLIC_KEY_ERROR` is really messing there, I added it. Other statuses just occur that APIML is not ready, so that can be handled via `sleep`